File: generate.js
import OpenAI from "openai";

const schema = {
  type: "object",
  additionalProperties: false,
  properties: {
    title: { type: "string" },
    subtitle: { type: "string" },
    slides: {
      type: "array",
      items: {
        type: "object",
        additionalProperties: false,
        properties: {
          title: { type: "string" },
          bullets: {
            type: "array",
            items: { type: "string" }
          },
          speaker_notes: { type: "string" },
          image_prompt: { type: "string" }
        },
        required: ["title", "bullets", "speaker_notes", "image_prompt"]
      }
    }
  },
  required: ["title", "subtitle", "slides"]
};

export default async function handler(req, res) {
  if (req.method === "OPTIONS") {
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");
    return res.status(204).end();
  }

  if (req.method !== "POST") {
    return res.status(405).json({ error: "Method not allowed." });
  }

  try {
    const { topic, slideCount, style, audience, language } = req.body || {};

    if (!topic || !String(topic).trim()) {
      return res.status(400).json({ error: "Please enter a presentation topic." });
    }

    const count = Number(slideCount || 10);
    if (!Number.isInteger(count) || count < 3 || count > 40) {
      return res.status(400).json({ error: "Slide count must be between 3 and 40." });
    }

    if (!process.env.OPENAI_API_KEY) {
      return res.status(500).json({
        error: "OPENAI_API_KEY is not configured in Vercel."
      });
    }

    const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

    const prompt = `
Create a professional PowerPoint presentation.

Topic: ${String(topic).trim()}
Exactly ${count} slides.
Style: ${style || "Professional"}
Audience: ${audience || "Students"}
Language: ${language || "English"}

Rules:
- Return exactly ${count} slides.
- Slide 1 should introduce the topic clearly.
- The final slide should contain key takeaways.
- Use concise, presentation-friendly wording.
- Prefer 3 to 6 bullets per content slide.
- Avoid paragraphs inside bullets.
- Keep facts accurate and suitable for the stated audience.
- Include useful speaker notes for every slide.
- For every slide, provide an image_prompt describing a relevant visual.
- Do not fabricate references or citations.
`;

    const response = await client.responses.create({
      model: process.env.OPENAI_MODEL || "gpt-5.6",
      input: prompt,
      text: {
        format: {
          type: "json_schema",
          name: "presentation_deck",
          strict: true,
          schema
        }
      }
    });

    const result = JSON.parse(response.output_text);

    if (!result.slides || result.slides.length !== count) {
      return res.status(502).json({
        error: "The AI returned an unexpected number of slides. Please try again."
      });
    }

    return res.status(200).json(result);
  } catch (error) {
    console.error(error);
    return res.status(500).json({
      error: error?.message || "Something went wrong while generating the presentation."
    });
  }
}

File: generate.py
import json,os
from http.server import BaseHTTPRequestHandler
from openai import OpenAI

SCHEMA={"type":"object","additionalProperties":False,"properties":{"slides":{"type":"array","items":{"type":"object","additionalProperties":False,"properties":{"title":{"type":"string"},"bullets":{"type":"array","items":{"type":"string"}},"speaker_notes":{"type":"string"},"image_prompt":{"type":"string"}},"required":["title","bullets","speaker_notes","image_prompt"]}}},"required":["slides"]}

def send(h,status,data):
 b=json.dumps(data).encode(); h.send_response(status); h.send_header("Content-Type","application/json"); h.send_header("Access-Control-Allow-Origin","*"); h.send_header("Access-Control-Allow-Headers","Content-Type"); h.send_header("Access-Control-Allow-Methods","POST, OPTIONS"); h.end_headers(); h.wfile.write(b)

class handler(BaseHTTPRequestHandler):
 def do_OPTIONS(self): send(self,200,{"ok":True})
 def do_POST(self):
  try:
   n=int(self.headers.get("Content-Length","0")); d=json.loads(self.rfile.read(n) or b"{}")
   topic=str(d.get("topic","")).strip(); count=int(d.get("slides",10)); style=d.get("style","Professional"); audience=d.get("audience","Students"); language=d.get("language","English")
   if not topic:return send(self,400,{"error":"Topic is required."})
   if not 3<=count<=40:return send(self,400,{"error":"Slides must be between 3 and 40."})
   key=os.getenv("OPENAI_API_KEY")
   if not key:return send(self,500,{"error":"OPENAI_API_KEY is not configured on the server."})
   client=OpenAI(api_key=key)
   prompt=f"""Create a polished PowerPoint presentation.
Topic: {topic}
Number of slides: exactly {count}
Style: {style}
Audience: {audience}
Language: {language}
Create exactly {count} slides. First slide should introduce the topic and final slide should summarize key takeaways. Use 3-6 concise bullets per slide. Keep information accurate and appropriate for the audience. For every slide include a useful image_prompt and concise speaker_notes. Do not invent citations."""
   r=client.responses.create(model=os.getenv("OPENAI_MODEL","gpt-5.6"),input=prompt,text={"format":{"type":"json_schema","name":"presentation_deck","strict":True,"schema":SCHEMA}})
   return send(self,200,json.loads(r.output_text))
  except Exception as e:return send(self,500,{"error":str(e)})
